src/transactions/cpfp.rs

```diff
@@ -0,0 +1,181 @@
+use crate::{
+    error::*,
+    transactions::{utils, CpfpableTransaction, RevaultTransaction, CPFP_MIN_CHANGE},
+    txins::*,
+    txouts::*,
+};
+
+use miniscript::bitcoin::{
+    consensus::encode::Decodable,
+    util::psbt::{Input as PsbtIn, PartiallySignedTransaction as Psbt},
+    Amount, Script, SigHashType, Transaction, TxIn, TxOut,
+};
+
+#[cfg(feature = "use-serde")]
+use {
+    serde::de::{self, Deserialize, Deserializer},
+    serde::ser::{Serialize, Serializer},
+};
+
+impl_revault_transaction!(
+    CpfpTransaction,
+    doc = "The transaction feebumping either an unvault or a spend"
+);
+impl CpfpTransaction {
+    /// Create a new cpfp transaction from the transaction that needs cpfp, its fees and weight;
+    /// the required feerate (expressed in sats/WU); a list of CpfpTxIn that can be used along
+    /// with their child number, the cpfp descriptor and the secp curve.
+    pub(crate) fn from_tx(
+        to_be_cpfped: &impl CpfpableTransaction,
+        feerate: f64,
+        mut listunspent: Vec<CpfpTxIn>,
+    ) -> Result<CpfpTransaction, TransactionCreationError> {
+        assert!(to_be_cpfped.is_finalized());
+        // Checking that the feerate requested is higher than the previous
+        // tx feerate (otherwise why cpfping in the first place?)
+        let to_be_cpfped_weight = to_be_cpfped.clone().into_tx().get_weight();
+        let to_be_cpfped_fees = Amount::from_sat(to_be_cpfped.fees());
+
+        if to_be_cpfped_fees.as_sat() as f64 / to_be_cpfped_weight as f64 >= feerate {
+            return Err(TransactionCreationError::FeerateTooLow);
+        }
+
+        // This will sort the vector in ascending order.
+        // Since we're going to pop() from it, we're using a largest first CS.
+        listunspent.sort_unstable_by_key(|l| l.txout().txout().value);
+
+        // Let's pick the input to be cpfped
+        let to_be_cpfped_index = listunspent
```

Comment:
I don't think the CPFP output should be computed externally and given as parameter: it really is a property of the "object" we are working on. You could have a `cpfp_txin()` method in the `CpfpTransaction` trait? [You already have it for the `UnvaultTransaction`](https://docs.rs/revault_tx/0.4.0/revault_tx/transactions/struct.UnvaultTransaction.html#method.cpfp_txin), could add it for the Spend.

Reply:
I'm not sure how to retrieve the derived_cpfp_descriptor from to_be_cpfped, and I need it for calling cpfp_txin. I don't think it's stored anywhere, is it?

I could ask for it in from_tx, but this makes the API rather ugly, no?